Improve sidecar injector documentation (#8927)
istio/istio.io: content/en/docs/setup/additional-setup/sidecar-injection/index.md

local-path-storage   Active   5m7s
{{< /text >}}

> Injection occurs at pod creation time. Kill the running pod and verify a new pod is created with the injected sidecar. The original pod has `2/2 READY` containers, and the pod with injected sidecar has `1/1 READY` containers.

Reviewer: ```suggestion
Injection occurs at pod creation time. Kill the running pod and verify a new pod is created with the injected sidecar. The original pod has `1/1 READY` containers, and the pod with injected sidecar has `2/2 READY` containers.
```

Author: Good catch, thanks!

sleep-776b7bcdcd-gmvnr   1/1       Running       0          2s
{{< /text >}}

#### Controlling the injection policy

In the above examples, you enabled and disabled injection at the namespace level. Injection can also be controlled
on a per-pod basis, by configuring the `sidecar.istio.io/inject` label on a pod:

| Resource | Label | Enabled value | Disabled value |
| -------- | ----- | ------------- | -------------- |
| Namespace | `istio-injection` | `enabled` | `disabled` |
| Pod | `sidecar.istio.io/inject` | `"true"` | `"false"` |

The injector is configured with the following logic:

1. If either label is disabled, the pod is not injected

> 1. If either label is enable, the pod is injected

Reviewer: ```suggestion
1. If either label is enabled, the pod is injected
```

> 1. If either label is enabled, the pod is injected

Reviewer 2: And no label is marked as disabled

Author: This was supposed to be an ordered list, with checks performed in order. I was wondering if it would not come off that way, apparently it does not :slightly_smiling_face:

Reviewer 2: Took me a while to get it :)

1. If neither label is set, the pod is injected if `.values.sidecarInjectorWebhook.enableNamespacesByDefault` is enabled. This is not enabled by default, so generally this means the pod is not injected.

## Customizing injection

Generally, pod are injected based on the sidecar injection template, configured in the `istio-sidecar-injector` configmap.

> Per-pod configuration is available to override these options on individual pods. This is done by adding a `istio-proxy` container

Reviewer: ```suggestion
Per-pod configuration is available to override these options on individual pods. This is done by adding an `istio-proxy` container
```

For example, the following configuration customizes a variety of settings, including lowering the CPU requests, adding a volume mount, and adding a `preStop` hook:

{{< text yaml >}}
apiVersion: v1
kind: Pod
metadata:
  name: example
spec:
  containers:
  - name: hello
    image: alpine
  - name: istio-proxy
    image: auto
    resources:
      requests:
        cpu: "100m"
    volumeMounts:
    - mountPath: /etc/certs
      name: certs
    lifecycle:
      preStop:
        exec:
          command: ["sleep", "10"]
  volumes:
  - name: certs
    secret:
      secretName: istio-certs
{{< /text >}}

In general, any field in a pod can be set. However, care must be taken for certain fields:

* Kubernetes requires the `image` field to be set before the injection has run. While you can set a specific image to override the default one,
  it is recommended to set the `image` to `auto` which will cause the sidecar injector to automatically select the image to use.
* Some fields in `Pod` are dependent on related settings. For example, CPU request must be less than CPU limit. If both fields are not configured together, the pod may fail to start.

> Additionally, certain fields are configurable by [annotations](/docs/reference/config/annotations/) on the pod, although it is recommend to use the above approach to customizing settings.

Reviewer 2: I think this is confusing.  Would not we recommend users to use annotations first, and if that doesn't provide the customization they need, they can look at the customization via pod yaml as described here.

Author: I don't see any reason to use the annotations now that we have the new way. WDYT?

Reviewer 2: I am not sure how to do proxy config annotation using the new way?  for example, if i want to use holdApplicationUntilProxyStarts.

Author: that one is an annotation :stuck_out_tongue: . I will include that in a new PR to clarify

### Custom templates [experimental]

{{< warning >}}

> This feature is experimental and subject to change, or be removed, at any time.

Reviewer: ```suggestion
This feature is experimental and subject to change, or removal, at any time.
```

{{< /warning >}}

Completely custom templates can also be defined at installation time.
For example, to define a custom template that injects the `GREETING` environment variable into the `istio-proxy` container:

{{< text yaml >}}
apiVersion: operator.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
spec:
  values:
    sidecarInjectorWebhook:
      templates:
        custom: |
          spec:
            containers:
            - name: istio-proxy
              env:
              - name: GREETING
                value: hello-world
{{< /text >}}

Pods will, by default, use the `sidecar` injection template, which is automatically created.
This can be overridden by the `inject.istio.io/templates` annotation.

> For example, to apply the default template and our customization, you can set `inject.istio.io/templates=sidecar,custom`.

Reviewer 2: sorry this example is confusing to me.  I'm not sure what the `sidecar,custom` mean here. would be nice to give out a consistent example as the name GREETING value hello-world earlier.

Author: There is a default `sidecar` template, which injects the sidecar. Additionally, user defines a `custom` template. They then tell the injector to use both of these. It will apply the sidecar template (injection istio-proxy container), then add the `custom` template (injecting GREETING env var)

Reviewer 2: I see, this is an annotation on the pod level to enable the default + custom injection template.
